Feature/add identity

the-man-w-laughs opened the pull request:
I have used:
IdentityServer4;
Clean architecture (at least tried);
Policy-based authentication;
Service result pattern;
Swagger;
Global exception handler;
Automapper.

the-man-w-laughs committed:
Add Identity.WebAPI.

the-man-w-laughs committed:
Add services to create users and gather token.

the-man-w-laughs committed:
bug/404-not-found

the-man-w-laughs committed:
Add SignatureValidator.

the-man-w-laughs committed:
Add signing credentials.

the-man-w-laughs committed:
Add services to create users and gather token.Merge fix/signing-key-bug to feature/add-identity.

the-man-w-laughs committed:
Add signing keys extensions to .gitignore

the-man-w-laughs committed:
Format solution.

katyalys commented on src/services/Identity/Identity.Application/ResultPattern/Results/InvalidResult.cs:
```diff
@@ -0,0 +1,18 @@
+    public class InvalidResult<T> : Result<T>
+    {
+        private string _error;
+        public InvalidResult(string error)
```
add empty line before constructor

katyalys commented on src/services/Identity/Identity.Application/Services/UserService.cs:
```diff
@@ -0,0 +1,95 @@
+        private readonly IMapper _mapper;
+        private readonly UserManager<AppUser> _userManager;
+        private readonly IAuthDBContext _authDbContext;
+
```
instead of IAuthDBContext implement repository pattern

katyalys commented on src/services/Identity/Identity.Application/Services/UserService.cs:
```diff
@@ -0,0 +1,95 @@
+            var appUser = _mapper.Map<AppUser>(appUserDto);
+
+            var identityResult = await _userManager.CreateAsync(appUser, appUserDto.Password);
+            if (!identityResult.Succeeded)
```
add empty line before if and before return

katyalys commented on src/services/Identity/Identity.Application/Services/UserService.cs:
```diff
@@ -0,0 +1,95 @@
+            if (!identityResult.Succeeded)
+            {                
+                var errorMessage = string.Join(", ", identityResult.Errors.Select(e => e.Description));
+                return new InvalidResult<AppUserDto>($"Unable to delete user: {errorMessage}");
```
empty line before return

katyalys commented on src/services/Identity/Identity.Infrastructure/DatabaseContext/AuthDbContext.cs:
```diff
@@ -0,0 +1,60 @@
+                    UserId = 1
+                });
+        }
+        public async Task<int> SaveChangesAsync()
```
SaveChangesAsync remove this method from dbContext class

katyalys commented on src/services/Identity/Identity.Infrastructure/DatabaseContext/AuthDbContext.cs:
```diff
@@ -0,0 +1,60 @@
+        });
+        }
+
+        private static void SetUsers(ModelBuilder builder)
```
create separate class for each of your set methods, and these classes should implement IEntityTypeConfiguration interface and then in OnModelCreating use method ApplyConfigurationsFromAssembly

the-man-w-laughs committed:
Format solution.

the-man-w-laughs committed:
Add IAppUserRepository

Major chanages:
- Add repository pattern;
- Remove SaveAsync in AuthDbContext.

the-man-w-laughs committed:
Add configurations to fill db.

Major changes:
- Add AddUserConfiguration, IdentityRoleConfiguration, IdentityUserRoleConfiguration;
- Apply configuration at OnModelCreating.

katyalys commented on src/services/Identity/Identity.Infrastructure/Repositories/AppUserRepository.cs:
```diff
@@ -0,0 +1,53 @@
+    private void HandleIdentityResult(string action, IdentityResult identityResult)
+    {
+        if (!identityResult.Succeeded)
+        {
```
repository should only interact with db and it doesn't check if there are errors, move error handling logic to services

the-man-w-laughs committed:
Move HandleIdentityResult from Repository to Service.

the-man-w-laughs committed:
Refactor UserService.
